### PR TITLE
Rewrite DB version 0020 to run online

### LINKED ISSUE #3235: Rewrite 0020 migration to be able to happen online
Migation 0020 took too long in the TCW this weekend.  We need to find a way to do this without downtime.

Rough outline from the team meeting just now: 
 * operate in a mode where we have both an "old" (queue_tasks_entities/queue_task_group_active_sets_entities) and "new" (tasks) table
 * provide some mechanism to migrate tasks one-by-one (or in small batches) from old to new

Parameters on the solution:
 * Online
 * Result must match the existing version 0020 exactly when finished (since we've already deployed 0020 elsewhere)
 * Optional: 
   * Support downgrades
     * Also online
       * If the upgrade only partially completed

### PR DESCRIPTION
**Pull request by djmitche**
Github Bug/Issue: Fixes #3235

### CODE REVIEW

File: db/versions/0020-downgrade.sql [hunk 1]
```diff
@@ -4,34 +4,46 @@ begin
   lock table tasks;
 
   raise log 'TIMING start queue_tasks_entities create table';
-  create table queue_tasks_entities(
+  create table if not exists queue_tasks_entities(
```
[the remark this answers is not in the document]
Comment (djmitche) [reply db/versions/0020-downgrade.sql:7]: it will already exist if the upward migration didn't finish -- this is accounting for downgrading in that situation

File: db/versions/0020-downgrade.sql [hunk 2]
```diff
@@ -4,34 +4,46 @@ begin
...
     value jsonb not null,
     version integer not null,
     etag uuid default public.gen_random_uuid());
 
   raise log 'TIMING start queue_tasks_entities primary key';
-  alter table queue_tasks_entities add primary key (partition_key, row_key);
+  begin
+    alter table queue_tasks_entities add primary key (partition_key, row_key);
+  exception
+    when SQLSTATE '42P16' then
```
[the remark this answers is not in the document]
Comment (djmitche) [reply db/versions/0020-downgrade.sql:17]: Huh, I just copy-pasted the code I got.  But, yes, probably?

File: db/versions/0020-migration.sql [hunk 3]
```diff
@@ -17,113 +11,577 @@ begin
...
-  drop table queue_task_group_members_entities;
-  revoke select, insert, update, delete on queue_task_group_active_sets_entities from $db_user_prefix$_queue;
-  drop table queue_task_group_active_sets_entities;
-
   grant select, insert, update, delete on tasks to $db_user_prefix$_queue;
+
+  -- define a function to march the migration forward (in a transaction) and
+  -- another to indicate the migration is complete.
+
+  create or replace function v20_migration_migrate_tasks(task_ids_after text, count int) returns setof text as $x$
```
[the remark this answers is not in the document]
Comment (djmitche) [reply db/versions/0020-migration.sql:54]: Exactly.  I couldn't think of a way to blend them into a single function, and anyway the second function is already gigantic.

**Comment by djmitche** [comment]
TODO:
 - lots more testing, including for corner cases like a task group defined partially before and partially after the initial script run
 - rebase onto a version without db 0021 and higher
 - write a temporary `yarn` subcommand to do this stuff
 - test this out in dev

**Comment by djmitche** [comment]
Slightly concerning: I actually forgot to rewrite most of the entities methods to look in both the old and new tables.. and yet the tests are passing.  So.  Lots more testing.

File: db/versions/0020.yml [hunk 4]
```diff
@@ -66,6 +77,12 @@ methods:
       declare
         new_row tasks%ROWTYPE;
       begin
+        -- create the task in the new table, checking for conflicts in the old
+        perform 1 from tasks where tasks.task_id = pk;
```
[the remark this answers is not in the document]
Comment (djmitche) [reply db/versions/0020.yml:81]: It should!  This got caught in a later test, but I'll double-check all of these to make sure they account for all keys and are looking at the right table.

File: db/versions/0020.yml [hunk 5]
```diff
@@ -109,10 +126,8 @@ methods:
     returns: table (etag uuid)
     body: |-
       begin
-        return query delete from tasks
-        where
-          tasks.task_id = partition_key
-        returning tasks.etag;
+        -- during this migration, removing tasks does nothing
```
[the remark this answers is not in the document]
Comment (djmitche) [reply db/versions/0020.yml:129]: Good eye -- further testing showed this was wrong, too.  I thought removals only occurred for expiration, and overall this tries to just "pause" expiration while the migration is running.  But it turns out we also remove tasks if, after adding them, we find an error such as an undefined dependency.  So I've modified this to only allow removals from the new table.

File: db/versions/0020.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (djmitche) [reply db/versions/0020.yml:81]: double-checked

**Comment by djmitche** [comment]
OK, then, I think this is ready to go.  I'm trying it out in my copy of production from a few weeks ago, while running the performance tester.  @sciurus do you want to try it out on a duplicate of the DB?  I can build a docker image from this, from which `docker run .. db:0020` in the same context as a "regular" DB upgrade should do the trick.  You can even ctrl-c and re-start it a few times.

**Comment by djmitche** [comment]
Docker image is `taskcluster/taskcluster:v34.0.0-3-ga13e82052`

**Comment by djmitche** [comment]
Updated docker image that uses the generated DB schema, since `db/versions` is removed from docker images -- `taskcluster/taskcluster:v34.0.0-3-g393053b0e` /cc @sciurus

**Comment by djmitche** [comment]
This is running at around 400 rows/s for me, with performance-tester running in parallel (which is creating new tasks as the migration proceeds).  @sciurus saw a slightly higher rate with no load on the DB.  Both look to finish in around 20 hours.  The performance tester has flagged a few minor issues (a DNS issue, for example) but nothing serious.

**Comment by djmitche** [comment]
Development here is done, and this was never intended to land, so I'm going to close this up.
